### PR TITLE
[Refactor] YOLOv6 BatchATSSAssigner

### PR DESCRIPTION
**Pull request by PeterH0323**
## Motivation

Refactor YOLOv6 BatchATSSAssigner and add UT

## Modification

`mmyolo/models/task_modules/assigners/batch_atss_assigner.py`

## Add

UT: `tests/test_models/test_task_modules/test_assigners/test_batch_atss_assigner.py`

image

### CODE REVIEW

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 1]
```diff
@@ -1,254 +1,339 @@
...
 import torch.nn.functional as F
+from mmdet.utils import ConfigType
+from torch import Tensor
 
 from mmyolo.registry import TASK_UTILS
 
-def dist_calculator(gt_bboxes, anchor_bboxes):
-    """compute center distance between all bbox and gt
+
+def bbox_center_distance(bboxes: Tensor, anchor: Tensor):
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:13]: anchor -> priors

Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:13]: nter_distance(bboxes: Tensor, anchor: Tensor): -> tuple

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 2]
```diff
@@ -1,254 +1,339 @@
...
-    
-    return distances, ac_points
-
-def select_candidates_in_gts(xy_centers, gt_bboxes, eps=1e-9):
-    """select the positive anchors's center in gt
+        bboxes (Tensor): Shape (n, 4) for , "xyxy" format.
+        anchor (Tensor): Shape (n, 4) for priors, "xyxy" format.
+
+    Returns:
+        Tensor: Center distances between bboxes and priors.
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:21]: out shape

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 3]
```diff
@@ -1,254 +1,339 @@
...
+    distances = (bbox_points[:, None, :] -
+                 anchor_points[None, :, :]).pow(2).sum(-1).sqrt()
+
+    return distances, anchor_points
+
+
+def select_candidates_in_gts(xy_centers: Tensor,
+                             gt_bboxes: Tensor,
+                             eps: float = 1e-9) -> Tensor:
+    """select the positive anchors' center in gt.
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:40]: anchor -> prior

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 4]
```diff
@@ -1,254 +1,339 @@
...
 
     Args:
         xy_centers (Tensor): shape(bs*n_max_boxes, num_total_anchors, 4)
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:43]: shape(bs*n_max_boxes, num_total_anchors, 4) -> shape (B*num_bboxes,num_anchors,4)

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 5]
```diff
@@ -1,254 +1,339 @@
...
-    target_gt_idx = mask_pos.argmax(axis=-2)
-    return target_gt_idx, fg_mask , mask_pos
 
-def iou_calculator(box1, box2, eps=1e-9):
-    """Calculate iou for batch
+    target_gt_index = mask_pos.argmax(axis=-2)
+    return target_gt_index, fg_mask, mask_pos
+
+
+def iou_calculator(bbox1: Tensor, bbox2: Tensor, eps: float = 1e-9) -> Tensor:
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:94]: Why not reuse mmdet？

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 6]
```diff
@@ -1,254 +1,339 @@
...
+    bbox_deltas = bbox_deltas.reshape(
+        [batch_size, max_boxes, anchors_number, -1])
+
     return (bbox_deltas.min(axis=-1)[0] > eps).to(gt_bboxes.dtype)
 
-def select_highest_overlaps(mask_pos, overlaps, n_max_boxes):
-    """if an anchor box is assigned to multiple gts,
-        the one with the highest iou will be selected.
+
+def select_highest_overlaps(mask_pos: Tensor, overlaps: Tensor,
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:65]: pos_mask

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 7]
```diff
@@ -1,254 +1,339 @@
...
+                            max_boxes: int) -> Tuple[Tensor, Tensor, Tensor]:
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:66]: max_boxes mean?

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 8]
```diff
@@ -1,254 +1,339 @@
...
     @torch.no_grad()
-    def forward(self,
-                anc_bboxes,
-                n_level_bboxes,
-                gt_labels,
-                gt_bboxes,
-                mask_gt,
-                pd_bboxes):
+    def forward(
+            self, anchor_bboxes: torch.Tensor, n_level_bboxes: List,
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:141]: anchor_bboxes -> priors

Comment (hhaAndroid) [reply mmyolo/models/task_modules/assigners/batch_atss_assigner.py:141]: n_level_bboxes -> num_level_priors

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 9]
```diff
@@ -1,254 +1,339 @@
...
+            gt_labels: torch.Tensor, gt_bboxes: torch.Tensor,
+            mask_gt: torch.Tensor,
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:143]: mask_gt mean?

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 10]
```diff
@@ -1,254 +1,339 @@
...
-            gt_labels (Tensor): shape(bs, n_max_boxes, 1)
-            gt_bboxes (Tensor): shape(bs, n_max_boxes, 4)
-            mask_gt (Tensor): shape(bs, n_max_boxes, 1)
-            pd_bboxes (Tensor): shape(bs, n_max_boxes, 4)
+            anchor_bboxes (Tensor): shape(num_total_anchors, 4)
+            n_level_bboxes (List): len(3)
+            gt_labels (Tensor): shape(batch_size, n_max_boxes, 1)
+            gt_bboxes (Tensor): shape(batch_size, n_max_boxes, 4)
+            mask_gt (Tensor): shape(batch_size, n_max_boxes, 1)
+            pd_bboxes (Tensor): shape(batch_size, n_max_boxes, 4)
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:154]: pd_bboxes mean?

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 11]
```diff
@@ -1,254 +1,339 @@
...
+            batch_size, max_boxes)
 
         # soft label with iou
         if pd_bboxes is not None:
             ious = iou_calculator(gt_bboxes, pd_bboxes) * mask_pos
             ious = ious.max(axis=-2)[0].unsqueeze(-1)
             target_scores *= ious
 
-        return target_labels.long(), target_bboxes, target_scores, fg_mask.bool()
+        return target_labels.long(
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:213]: For extensibility, the return type is preferably dict

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 12]
```diff
@@ -1,254 +1,398 @@
...
+        target_gt_index (Tensor): Target ground true index,
+            shape(batch_size, num_priors)
+        force_gt_matching (Tensor): Force matching ground true,
+            shape(batch_size, num_priors)
+        possible_mask (Tensor): Possible mask,
+            shape(batch_size, num_gt, num_priors)
+    """
+    force_gt_matching = possible_mask.sum(axis=-2)
+
+    # make sure all gt_bbox match anchor
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:95]:
```suggestion
    # Make sure the positive sample matches the only one and is the largest IoU
```

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 13]
```diff
@@ -1,254 +1,398 @@
...
+        index = overlaps.argmax(axis=1)
+        is_max_overlaps = F.one_hot(index, num_gt)
+        is_max_overlaps = \
+            is_max_overlaps.permute(0, 2, 1).to(overlaps.dtype)
+
+        possible_mask = torch.where(mask_multi_gts, is_max_overlaps,
+                                    possible_mask)
+        force_gt_matching = possible_mask.sum(axis=-2)
+
+    target_gt_index = possible_mask.argmax(axis=-2)
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:108]: target_gt_index -> gt_idx_pre_prior

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py (hunk 5 above)
Comment (hhaAndroid) [reply mmyolo/models/task_modules/assigners/batch_atss_assigner.py:112]: The result should be the same

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 14]
```diff
@@ -1,254 +1,398 @@
...
-                anc_bboxes,
-                n_level_bboxes,
-                gt_labels,
-                gt_bboxes,
-                mask_gt,
-                pd_bboxes):
+    def forward(
+            self, priors: torch.Tensor, num_level_priors: List,
+            gt_labels: torch.Tensor, gt_bboxes: torch.Tensor,
+            pad_gt_mask: torch.Tensor, pred_bboxes: torch.Tensor
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:160]: pad_gt_mask -> pad_bbox_flag

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 15]
```diff
@@ -1,254 +1,398 @@
...
+    ) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
         r"""This code is based on
             https://github.com/fcjian/TOOD/blob/master/mmdet/core/bbox/assigners/atss_assigner.py
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:163]: replace to yolov6 atssassigner link

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 16]
```diff
@@ -1,254 +1,398 @@
...
-                   torch.zeros([self.bs, self.n_anchors]).to(device)
-
-
-        overlaps = self.iou2d_calculator(gt_bboxes.reshape([-1, 4]), anc_bboxes)
-        overlaps = overlaps.reshape([self.bs, -1, self.n_anchors])
-
-        distances, ac_points = dist_calculator(gt_bboxes.reshape([-1, 4]), anc_bboxes)
-        distances = distances.reshape([self.bs, -1, self.n_anchors])
-
+            return (torch.full([batch_size, num_priors],
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:190]: return gt_bboxes.new_full().   Do not use the usage of to(device)

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 17]
```diff
@@ -1,254 +1,398 @@
...
+            is_in_candidate, candidate_idxs, overlaps, num_priors, batch_size,
+            num_gt)
 
-        overlaps_thr_per_gt, iou_candidates = self.thres_calculator(
-            is_in_candidate, candidate_idxs, overlaps)
-        
         # select candidates iou >= threshold as positive
-        is_pos = torch.where(
-            iou_candidates > overlaps_thr_per_gt.repeat([1, 1, self.n_anchors]),
+        is_possible = torch.where(
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:217]: is_possiable -> is_pos

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 18]
```diff
@@ -1,254 +1,398 @@
...
+            iou_candidates > overlaps_thr_per_gt.repeat([1, 1, num_priors]),
             is_in_candidate, torch.zeros_like(is_in_candidate))
 
-        is_in_gts = select_candidates_in_gts(ac_points, gt_bboxes)
-        mask_pos = is_pos * is_in_gts * mask_gt
+        is_in_gts = select_candidates_in_gts(priors_points, gt_bboxes)
+        possible_mask = is_possible * is_in_gts * pad_gt_mask
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:222]: possable_mask -> pos_mask

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 19]
```diff
@@ -1,254 +1,398 @@
...
+
 @TASK_UTILS.register_module()
 class BatchATSSAssigner(nn.Module):
-    '''Adaptive Training Sample Selection Assigner'''
-    def __init__(self,
-                 topk=9,
-                 iou2d_calculator=dict(type='mmdet.BboxOverlaps2D'),
-                 num_classes=80):
-        super(BatchATSSAssigner, self).__init__()
+    """Adaptive Training Sample Selection Assigner."""
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:144]: Add link to yolov6 assigner

Comment (hhaAndroid) [reply mmyolo/models/task_modules/assigners/batch_atss_assigner.py:144]: Need to add unit tests

File: mmyolo/models/task_modules/assigners/batch_task_aligned_assigner.py [hunk 20]
```diff
@@ -2,14 +2,15 @@
...
-from .batch_atss_assigner import select_candidates_in_gts, select_highest_overlaps, iou_calculator
+from .batch_atss_assigner import select_candidates_in_gts, select_highest_overlaps
 from mmyolo.registry import TASK_UTILS
 
 
 @TASK_UTILS.register_module()
 class BatchTaskAlignedAssigner(nn.Module):
     def __init__(self, 
                  topk=13,
+                 iou_calculator=None,
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_task_aligned_assigner.py:13]: `num_classes` is placed at the front and does not provide a default value, `iou_calculator` should provide a default value

File: tests/test_models/test_task_modules/test_assigners/test_batch_atss_assigner.py [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+
+
+class TestBatchATSSAssigner(TestCase):
+
+    def test_batch_atss_assigner(self):
+        batch_atss_assigner = BatchATSSAssigner(
+            topk=9,
+            iou_calculator=dict(type='mmdet.BboxOverlaps2D'),
+            num_classes=80)
+        batch_size = 16
```
Comment (hhaAndroid) [inline tests/test_models/test_task_modules/test_assigners/test_batch_atss_assigner.py:16]: batch_size = 2

File: tests/test_models/test_task_modules/test_assigners/test_batch_atss_assigner.py [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+from mmyolo.models.task_modules.assigners import BatchATSSAssigner
+
+
+class TestBatchATSSAssigner(TestCase):
+
+    def test_batch_atss_assigner(self):
+        batch_atss_assigner = BatchATSSAssigner(
+            topk=9,
+            iou_calculator=dict(type='mmdet.BboxOverlaps2D'),
+            num_classes=80)
```
Comment (hhaAndroid) [inline tests/test_models/test_task_modules/test_assigners/test_batch_atss_assigner.py:15]: num_classes=2

File: tests/test_models/test_task_modules/test_assigners/test_batch_atss_assigner.py [hunk 23]
```diff
@@ -0,0 +1,135 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from unittest import TestCase
+
+import torch
+
+from mmyolo.models.task_modules.assigners import BatchATSSAssigner
+
+
+class TestBatchATSSAssigner(TestCase):
```
Comment (hhaAndroid) [inline tests/test_models/test_task_modules/test_assigners/test_batch_atss_assigner.py:9]: Unit tests should use fewer resources. so many parameters can be changed to smaller

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 24]
```diff
@@ -1,254 +1,368 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+from typing import List, Tuple
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:2]: yolov6 is an anchor-free algorithm, so it is best not to have the anchor field and should use prior uniformly

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 25]
```diff
@@ -1,254 +1,368 @@
...
-    """Calculate iou for batch
+def select_highest_overlaps(pos_mask: Tensor, overlaps: Tensor,
+                            num_gt: int) -> Tuple[Tensor, Tensor, Tensor]:
+    """If an anchor box is assigned to multiple gts, the one with the highest
+    iou will be selected.
 
     Args:
-        box1 (Tensor): shape(bs, n_max_boxes, 1, 4)
-        box2 (Tensor): shape(bs, 1, num_total_anchors, 4)
+        pos_mask (Tensor): Possible mask,
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:80]: Possible mask -> The assigned positive sample mask

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 26]
```diff
@@ -1,254 +1,368 @@
...
+        overlaps (Tensor): IoU between all bbox and ground true,
+            shape(batch_size, num_gt, num_priors)
+        num_gt (int): Number of ground true.
     Return:
-        (Tensor): shape(bs, n_max_boxes, num_total_anchors)
+        gt_idx_pre_prior (Tensor): Target ground true index,
+            shape(batch_size, num_priors)
+        fg_mask_pre_prior (Tensor): Force matching ground true,
+            shape(batch_size, num_priors)
+        pos_mask (Tensor): Possible mask,
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:90]: same as above

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 27]
```diff
@@ -1,254 +1,368 @@
...
+
+    This code is based on
+    https://github.com/meituan/YOLOv6/blob/main/yolov6/assigners/atss_assigner.py
+    """
+
+    def __init__(
+            self,
+            topk: int = 9,
+            iou_calculator: ConfigType = dict(type='mmdet.BboxOverlaps2D'),
+            num_classes: int = 80):
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:123]: num_classes cannot set default value

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 28]
```diff
@@ -1,254 +1,368 @@
...
+    ) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
+        r"""Get assigner result
         Args:
-            anc_bboxes (Tensor): shape(num_total_anchors, 4)
-            n_level_bboxes (List):len(3)
-            gt_labels (Tensor): shape(bs, n_max_boxes, 1)
-            gt_bboxes (Tensor): shape(bs, n_max_boxes, 4)
-            mask_gt (Tensor): shape(bs, n_max_boxes, 1)
-            pd_bboxes (Tensor): shape(bs, n_max_boxes, 4)
+            priors (Tensor): model predictions, it can be anchors, points,
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:137]: model predictions？

File: mmyolo/models/task_modules/assigners/batch_atss_assigner.py [hunk 29]
```diff
@@ -1,254 +1,368 @@
...
-                gt_bboxes,
-                mask_gt,
-                pd_bboxes):
-        r"""This code is based on
-            https://github.com/fcjian/TOOD/blob/master/mmdet/core/bbox/assigners/atss_assigner.py
-
+    def forward(
+            self, priors: torch.Tensor, num_level_priors: List,
+            gt_labels: torch.Tensor, gt_bboxes: torch.Tensor,
+            pad_bbox_flag: torch.Tensor, pred_bboxes: torch.Tensor
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_atss_assigner.py:133]: pred_bboxes： Option[tensor]=None

File: mmyolo/models/task_modules/assigners/batch_task_aligned_assigner.py (hunk 20 above)
Comment (PeterH0323) [reply mmyolo/models/task_modules/assigners/batch_task_aligned_assigner.py:13]: 这个文件是还没重构的，会在另外一个 PR 重构

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**